src/ServiceControl.Transports.AmazonSQS/QueueLengthProvider.cs

```diff
@@ -0,0 +1,132 @@
+﻿namespace ServiceControl.Transports.AmazonSQS
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Amazon.SQS;
+    using Amazon.SQS.Model;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.Messaging;
+    using Monitoring.QueueLength;
+    using NServiceBus.Logging;
+    using NServiceBus.Metrics;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        ConcurrentDictionary<EndpointInputQueue, string> queues = new ConcurrentDictionary<EndpointInputQueue, string>();
+        ConcurrentDictionary<string, int> sizes = new ConcurrentDictionary<string, int>();
+
+        QueueLengthStore store;
+
+        CancellationTokenSource stop = new CancellationTokenSource();
+        Task pooler;
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            this.store = store;
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            var endpointInputQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+            var queue = QueueNameHelper.GetSqsQueueName(metadataReport.LocalAddress);
+
+            queues.AddOrUpdate(endpointInputQueue, _ => queue, (_, currentQueue) =>
+            {
+                if (currentQueue != queue)
+                {
+                    sizes.TryRemove(currentQueue, out var _);
+                }
+
+                return queue;
+            });
+
+            sizes.TryAdd(queue, 0);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+            //HINT: ASQ  server endpoints do not support endpoint level queue length monitoring
+        }
+
+        public Task Start()
+        {
+            stop = new CancellationTokenSource();
+
+            pooler = Task.Run(async () =>
+            {
+                using (var client = new AmazonSQSClient())
+                {
+                    var cache = new QueueUrlCache(client);
+
+                    while (!stop.Token.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await FetchQueueSizes(cache, client).ConfigureAwait(false);
+
+                            UpdateQueueLengthStore();
+
+                            await Task.Delay(QueryDelayInterval);
+                        }
+                        catch (Exception e)
+                        {
+                            Logger.Error("Error querying sql queue sizes.", e);
```

Comment:
`Error querying sql queue sizes.` should be `SQS` and probably give a hint about what went wrong?